kvpb: remove RequestHeaderPure

RaduBerinde opened the pull request:
*Note: this PR is only for the top commit.*

Remove `RequstHeaderPure/RequestHeaderCrdbTest` and use
`gogoproto.omitempty` instead.

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/135659)
<!-- Reviewable:end -->

tbg reviewed (approved):
> Reviewed 5 of 5 files at r1, 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135659)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi)



<!-- Sent from Reviewable.io -->

RaduBerinde committed:
kvpb: remove RequestHeaderPure

Remove `RequstHeaderPure/RequestHeaderCrdbTest` and use
`gogoproto.omitempty` instead.

Epic: none
Release note: None
